fix: Fix SQLite issue

dmga44 opened the pull request:
An issue appeared when several processes/threads started doing
concurrent operations on a SQLite database in our system. The issue originated from
having different ReadRecords for each those processes while also doing
writes on the same file the cache was about. Solution is to clear the
cache after chunk is locked and prepared to be written.

A couple of tests were added to test the fix.

Signed-off-by: Dave <dave@leil.io>
Co-authored-by: aNeutrino <pm@leil.io>

dmga44 committed:
fix: Fix SQLite issue

An issue appeared when several processes/threads started doing
concurrent operations an SQLite database. The issue originated from
having different ReadRecords for each those processes while also doing
writes on the same file the cache was about. Solution is to clear the
cache after chunk is locked and prepared to be written.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
test: Test SQLite fix

A couple of new tests were added to the ShortSystemTests and
LongSystemTests suites. These tests intend to check every time that the
SQLite workload is supported by our system.

Signed-off-by: Dave <dave@leil.io>

Co-authored-by: aNeutrino <pm@leil.io>

dmga44 commented on src/mount/readdata.cc:
```diff
@@ -672,6 +672,21 @@ void read_inode_ops(uint32_t inode) { // attributes of inode have been changed -
+			it->second->cache.selective_clear(chunkId);  // clear cache for this chunk
+		}
+	}
+}
```
The mutexes are nested because each of them is required for protection of the structures they protect (`gMutex` protects the `gActiveReadRecords` and the `inodeLock` protects the cache). The issue is that in some other places in the code the `ReadRecord` were are iterating could be released, so if the `gMutex` is not locked then we would have SEGFAULT if we follow your proposal.

dmga44 commented on src/mount/readdata_cache.h:
```diff
@@ -410,6 +410,20 @@ class ReadCache {
 		}
 	}
 
+	void selective_clear(uint32_t chunkId) {
+		uint64_t chunkStart = chunkId * SFSCHUNKSIZE;
+		uint64_t chunkEnd = chunkStart + SFSCHUNKSIZE;
+		auto it = entries_.begin();
+		while (it != entries_.end()) {
+			// If the entry is not in the chunk, skip it
+			if (it->endOffset() < chunkStart || it->offset >= chunkEnd) {
+				++it;
+				continue;
+			}
+			it = erase(it);
+		}
+	}
```
The erase makes sure that the return value is the next iterator.

lgsilva3087 commented on src/mount/readdata_cache.h:
```diff
@@ -410,6 +410,20 @@ class ReadCache {
 		}
 	}
 
+	void selective_clear(uint32_t chunkId) {
```
Consider chunkIndex instead of chunkId, which is used for other purposes.

dmga44 replied:
Indeed, renamed, thanks.

lgsilva3087 commented on src/mount/readdata_cache.h:
```diff
@@ -410,6 +410,20 @@ class ReadCache {
+	void selective_clear(uint32_t chunkId) {
+		uint64_t chunkStart = chunkId * SFSCHUNKSIZE;
+		uint64_t chunkEnd = chunkStart + SFSCHUNKSIZE;
+		auto it = entries_.begin();
```
I don't know the expected number of entries, if needed, please consider:
- Using lower_bound to go directly to the interesting ones (assuming the intrusive set keeps them sorted).
- Divide the condition to avoid iterating over the last entries after the chunkEnd:
```
if (it->offset >= chunkEnd) {
	break;
}
```

dmga44 replied:
I thought that in the end, these kind of workloads where several threads write and read though the same mountpoint to the same file would perform bad anyway and this would be the less impacting part, but I get your point it would be nice not to waste time here.

I've applied the suggestions to make it iterate over the possible set of entries that could overlap the target chunk. Thanks.

lgsilva3087 commented on src/mount/readdata.h:
```diff
@@ -382,6 +382,7 @@ uint32_t read_data_get_total_read_timeout_ms();
 bool read_data_get_prefetchxorstripes();
 
 void read_inode_ops(uint32_t inode);
+void read_inode_reconnect_clear(uint32_t inode, uint32_t chunkId);
```
Consider renaming it to: read_inode_reconnect_and_clear_cache.

dmga44 replied:
Done.

lgsilva3087 reviewed (approved):
Looks good, please see my suggestions.

rolysr commented on tests/test_suites/LongSystemTests/test_sqlite_stress_test_long.sh:
```diff
@@ -0,0 +1,26 @@
+timeout_set 1 hour
```
Consider to check the total time execution for this test to be as long as the other from this same test suite.

dmga44 replied:
I reduced the amount of ops it does in total, I also reduced the expected running time.

dmga44 committed:
refactor: Apply PR suggestions

Changes:
- rename parameter of read_inode_reconnect_and_clear_cache and
selective_clear align it to its true meaning.
- make the selective_clear function iterate over less elements.
- reduce the runtime and expected runtime of the
test_sqlite_stress_test_long.

Signed-off-by: Dave <dave@leil.io>

Co-authored-by: guillex <guillex@leil.io>
Co-authored-by: Rolando Sánchez Ramos <rolysr@leil.io>
